runtime/src/cost_tracker.rs

```diff
@@ -242,6 +251,23 @@ impl CostTracker {
             .saturating_sub(tx_cost.account_data_size);
         self.transaction_count = self.transaction_count.saturating_sub(1);
     }
+
+    fn adjust_transaction_execution_cost(&mut self, tx_cost: &TransactionCost, adjustment: i128) {
+        for account_key in tx_cost.writable_accounts.iter() {
+            let account_cost = self
+                .cost_by_writable_accounts
+                .entry(*account_key)
+                .or_insert(0);
+            *account_cost = u64::try_from(i128::from(*account_cost).saturating_add(adjustment))
+                .unwrap_or(*account_cost);
+        }
+        self.block_cost = u64::try_from(i128::from(self.block_cost).saturating_add(adjustment))
+            .unwrap_or(self.block_cost);
+        if tx_cost.is_simple_vote {
+            self.vote_cost = u64::try_from(i128::from(self.vote_cost).saturating_add(adjustment))
+                .unwrap_or(self.vote_cost);
```

Comment:
Oof. Yeah, either make cost signed, or make a function to add i64 to u64, like in https://doc.rust-lang.org/std/primitive.u64.html#method.saturating_add_signed, implemented in https://github.com/rust-lang/rust/blob/master/library/core/src/num/uint_macros.rs#L1064

(the adjustment could be checked to fit i64 once)

Reply:
Not super comfortable with cost being signed  - it makes sense that _adjustment_.cost has sign, but not so much of saying a transaction has `-100` units. 

Another option is to define adjustment as struct:
```
struct CostAdjustment {
    units: u64,
    adjust_op: enum{ Up, Down},
}

impl CostAdjustment {
    fn apply_to(&self, &mut cost) {
        match self.adjust_op {
            Up => cost.saturating_add(self.units),
            Down => cost.saturating_sub(self.units),
        }
    }
}
```
so the use site code would be: 
```
let adjustment = if lhs>rhs { CostAdjustment{ units = lhs-rhs, adjust_op = Down, } else {...};
adjustment.apply_to(lhs);
```
I don't know  if I like this any better after all. Open for ideas, maybe just take `u64.saturating_signe_add(i64)` approach?

Reply:
I think we should copy and use the `checked_add_signed` implementation from here https://github.com/rust-lang/rust/pull/87601/files. And then handle the case where `None` is returned as @t-nelson is suggesting

Reply:
> I think i64 is plenty here, no need for the i128 hit

Agreed, the rust-lang impl does exactly that

Reply:
Sounds good. As for handling overflow caused by adjustment, what do you think to immediately bump cost_tracker to its up limits to prevent writing more transactions into current block, so avoid making large blocks?

Reply:
[commit](f1371e2b7e0c35a1dcef0d98ea4cab3bf7367dd5) for math.